src/main/java/bio/terra/folder/service/create/CreateService.java

```diff
@@ -51,4 +40,60 @@ public String createFolder(CreateFolderBody folderBody, AuthenticatedUserRequest
     jobBuilder.submit();
     return folderId;
   }
+
+  private void validateRequest(CreateFolderBody request) {
+    String folderName = request.getName();
+    String parentFolderId = request.getParentFolderId().orElse(null);
+    // Require names only include alphanumeric characters, spaces, - and _
+    if (!folderName.matches("^[\\w\\-\\s_]+$")) {
+      throw new InvalidNameException(
+          "Provided name "
+              + folderName
+              + " contains invalid character(s). Valid names must only include alphanumeric"
+              + " characters, spaces, -, and _ characters, and cannot be empty.");
+    }
```

Comment:
for perf reasons, can you precompile the `^[\\w\\-\\s_]+$` regex as a `Pattern` and store it in a constant somewhere  (hey, it might be reusable elsewhere), then call `PRECOMPILED_FOLDER_PATTERN.matcher(folderName).matches()`?

See https://www.baeldung.com/java-regex-pre-compile for details ... precompiling regexes optimizes object creation, etc.

Also, is it possible for `folderName` to be null? I think not, from reading the code.

Reply:
Thanks for the article, good to know String.matches() was actually just creating new patterns. I left the pattern as static inside `CreateService` for now, but I can move it to a util if/when something else needs it.

re: `folderName` I didn't think so, but it looks like I forgot to mark the `name` request field as required in the API, so you actually could get a null pointer exception + 500 response. Added that in + another test, now it's a 400 with a useful error message.